File: internal/domain/item.go
package domain

import (
	"time"
)

// ItemStatus represents the status of an item
type ItemStatus string

const (
	ItemStatusAvailable   ItemStatus = "available"
	ItemStatusPawned      ItemStatus = "pawned"
	ItemStatusCollateral  ItemStatus = "collateral"
	ItemStatusForSale     ItemStatus = "for_sale"
	ItemStatusSold        ItemStatus = "sold"
	ItemStatusConfiscated ItemStatus = "confiscated"
	ItemStatusTransferred ItemStatus = "transferred"
	ItemStatusInTransfer  ItemStatus = "in_transfer"
	ItemStatusDamaged     ItemStatus = "damaged"
	ItemStatusLost        ItemStatus = "lost"
)

// ItemCondition represents the condition of an item
type ItemCondition string

const (
	ItemConditionNew       ItemCondition = "new"
	ItemConditionExcellent ItemCondition = "excellent"
	ItemConditionGood      ItemCondition = "good"
	ItemConditionFair      ItemCondition = "fair"
	ItemConditionPoor      ItemCondition = "poor"
)

// Item represents a pawn item
type Item struct {
	ID         int64  `json:"id"`
	BranchID   int64  `json:"branch_id"`
	CategoryID *int64 `json:"category_id,omitempty"`
	CustomerID *int64 `json:"customer_id,omitempty"` // Original owner

	// Identification
	SKU          string  `json:"sku"`
	Name         string  `json:"name"`
	Description  *string `json:"description,omitempty"`
	Brand        *string `json:"brand,omitempty"`
	Model        *string `json:"model,omitempty"`
	SerialNumber *string `json:"serial_number,omitempty"`
	Color        *string `json:"color,omitempty"`
	Condition    string  `json:"condition"`

	// Valuation
	AppraisedValue float64  `json:"appraised_value"` // Market value
	LoanValue      float64  `json:"loan_value"`      // Maximum loan amount
	SalePrice      *float64 `json:"sale_price,omitempty"`

	// Status
	Status ItemStatus `json:"status"`

	// Physical details (for jewelry)
	Weight float64  `json:"weight,omitempty"` // In grams
	Purity *string  `json:"purity,omitempty"` // e.g., 14k, 18k, .925

	// Additional info
	Notes *string  `json:"notes,omitempty"`
	Tags  []string `json:"tags,omitempty"`

	// Acquisition info
	AcquisitionType  string     `json:"acquisition_type"` // pawn, purchase, confiscation
	AcquisitionDate  time.Time  `json:"acquisition_date"`
	AcquisitionPrice *float64   `json:"acquisition_price,omitempty"`

	// Media
	Photos []string `json:"photos,omitempty"`

	// Audit
	CreatedBy int64 `json:"created_by,omitempty"`
	UpdatedBy int64 `json:"updated_by,omitempty"`

	// Timestamps
	CreatedAt time.Time  `json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`
	DeletedAt *time.Time `json:"deleted_at,omitempty"`

	// Relations
	Branch   *Branch   `json:"branch,omitempty"`
	Category *Category `json:"category,omitempty"`
	Customer *Customer `json:"customer,omitempty"`
}

// TableName returns the database table name
func (Item) TableName() string {
	return "items"
}

// IsAvailable checks if the item is available for loan or sale
func (i *Item) IsAvailable() bool {
	return i.Status == ItemStatusAvailable
}

// CanBeSold checks if the item can be sold
func (i *Item) CanBeSold() bool {
	return i.Status == ItemStatusAvailable || i.Status == ItemStatusConfiscated
}

// Acquisition type constants
const (
	AcquisitionTypePawn         = "pawn"
	AcquisitionTypePurchase     = "purchase"
	AcquisitionTypeConfiscation = "confiscation"
)

// ItemHistory represents a history entry for an item
type ItemHistory struct {
	ID            int64     `json:"id"`
	ItemID        int64     `json:"item_id"`
	Action        string    `json:"action"`
	OldStatus     string    `json:"old_status,omitempty"`
	NewStatus     string    `json:"new_status,omitempty"`
	OldBranchID   *int64    `json:"old_branch_id,omitempty"`
	NewBranchID   *int64    `json:"new_branch_id,omitempty"`
	ReferenceType *string   `json:"reference_type,omitempty"`
	ReferenceID   *int64    `json:"reference_id,omitempty"`
	Notes         string    `json:"notes,omitempty"`
	CreatedBy     int64     `json:"created_by,omitempty"`
	CreatedAt     time.Time `json:"created_at"`
}

// TableName returns the database table name
func (ItemHistory) TableName() string {
	return "item_history"
}

File: web/src/types/item.ts
// Item types - mirrors internal/domain/item.go

import type { Branch } from './branch'
import type { Category } from './category'
import type { Customer } from './customer'

export type ItemStatus =
  | 'available'
  | 'pawned'
  | 'collateral'
  | 'for_sale'
  | 'sold'
  | 'confiscated'
  | 'transferred'
  | 'in_transfer'
  | 'damaged'
  | 'lost'

export type ItemCondition = 'new' | 'excellent' | 'good' | 'fair' | 'poor'
export type AcquisitionType = 'pawn' | 'purchase' | 'consignment'

export const ITEM_STATUSES: { value: ItemStatus; label: string; color: string }[] = [
  { value: 'available', label: 'Disponible', color: 'green' },
  { value: 'pawned', label: 'Empeñado', color: 'blue' },
  { value: 'collateral', label: 'En Garantía', color: 'purple' },
  { value: 'for_sale', label: 'En Venta', color: 'orange' },
  { value: 'sold', label: 'Vendido', color: 'gray' },
  { value: 'confiscated', label: 'Confiscado', color: 'red' },
  { value: 'transferred', label: 'Transferido', color: 'cyan' },
  { value: 'in_transfer', label: 'En Tránsito', color: 'yellow' },
  { value: 'damaged', label: 'Dañado', color: 'red' },
  { value: 'lost', label: 'Perdido', color: 'red' },
]

export const ITEM_CONDITIONS: { value: ItemCondition; label: string }[] = [
  { value: 'new', label: 'Nuevo' },
  { value: 'excellent', label: 'Excelente' },
  { value: 'good', label: 'Bueno' },
  { value: 'fair', label: 'Regular' },
  { value: 'poor', label: 'Malo' },
]

export interface Item {
  id: number
  branch_id: number
  category_id?: number
  customer_id?: number

  // Identification
  sku: string
  name: string
  description?: string
  brand?: string
  model?: string
  serial_number?: string
  color?: string
  condition: ItemCondition

  // Valuation
  appraised_value: number
  loan_value: number
  sale_price?: number

  // Status
  status: ItemStatus

  // Physical details
  weight?: number
  purity?: string

  // Additional info
  notes?: string
  tags?: string[]

  // Acquisition
  acquisition_type: AcquisitionType
  acquisition_date: string
  acquisition_price?: number

  // Media
  photos?: string[]

  // Delivery tracking
  delivered_at?: string

  // Audit
  created_by?: number
  updated_by?: number
  created_at: string
  updated_at: string
  deleted_at?: string

  // Relations
  branch?: Branch
  category?: Category
  customer?: Customer
}

export interface CreateItemInput {
  branch_id: number
  category_id?: number
  customer_id?: number
  name: string
  description?: string
  brand?: string
  model?: string
  serial_number?: string
  color?: string
  condition: ItemCondition
  appraised_value: number
  loan_value: number
  sale_price?: number
  weight?: number
  purity?: string
  notes?: string
  tags?: string[]
  acquisition_type: AcquisitionType
  acquisition_price?: number
}

export interface UpdateItemInput {
  category_id?: number
  name?: string
  description?: string
  brand?: string
  model?: string
  serial_number?: string
  color?: string
  condition?: ItemCondition
  appraised_value?: number
  loan_value?: number
  sale_price?: number
  weight?: number
  purity?: string
  notes?: string
  tags?: string[]
}

export interface ItemListParams {
  page?: number
  per_page?: number
  order_by?: string
  order?: 'asc' | 'desc'
  search?: string
  branch_id?: number
  category_id?: number
  customer_id?: number
  status?: ItemStatus
  condition?: ItemCondition
}

export interface ItemHistory {
  id: number
  item_id: number
  action: string
  old_status?: string
  new_status?: string
  old_branch_id?: number
  new_branch_id?: number
  reference_type?: string
  reference_id?: number
  notes?: string
  created_by?: number
  created_at: string
}
